### PR TITLE
Implement multi-stage Docker build

### PR DESCRIPTION
**Pull request by rgreinho**
# Description

Updates the Dockerfile to leverage multi-stage build and ensure that the
runtime contains only the dependencies required to run OpsDroid.

The final image was updated to run as an unprivileged user.

The concepts used to build this new version of the Docker image mostly
come from Bitnami's recommendations:
https://engineering.bitnami.com/articles/best-practices-writing-a-dockerfile.html

The wheel `unversal` flag was removed since the project does not have C
extensions, nor supports Python 2 (see
https://packaging.python.org/guides/distributing-packages-using-setuptools/#wheels
for reference).

## Status
**READY**

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Building and running the image locally.

```
$ docker run --rm -it opsdroid/opsdroid:snapshot
INFO opsdroid.logging: ========================================
INFO opsdroid.logging: Started opsdroid 0+unknown.
INFO opsdroid: ========================================
INFO opsdroid: You can customise your opsdroid by modifying your configuration.yaml.
INFO opsdroid: Read more at: http://opsdroid.readthedocs.io/#configuration
INFO opsdroid: Watch the Get Started Videos at: http://bit.ly/2fnC0Fh
INFO opsdroid: Install Opsdroid Desktop at:
https://github.com/opsdroid/opsdroid-desktop/releases
INFO opsdroid: ========================================
WARNING opsdroid.loader: No databases in configuration. This will cause skills which store things in memory to lose data when opsdroid is restarted.
INFO opsdroid.loader: Cloning dance from remote repository.
INFO opsdroid.loader: Cloning hello from remote repository.
INFO opsdroid.loader: Cloning loudnoises from remote repository.
INFO opsdroid.loader: Cloning seen from remote repository.
INFO opsdroid.core: Opsdroid is now running, press ctrl+c to exit.
INFO opsdroid.web: Started web server on http://0.0.0.0:8080
^CINFO opsdroid.core: Received stop signal, exiting.
INFO opsdroid.core: Stopping connector websocket...
INFO opsdroid.core: Stopped connector websocket.
INFO opsdroid.core: Stopping database inmem...
INFO opsdroid.core: Stopped database inmem.
INFO opsdroid.core: Stopping web server...
INFO opsdroid.core: Stopped web server.
INFO opsdroid.core: Stopping pending tasks...
INFO opsdroid.core: Stopped pending tasks.
INFO opsdroid.core: Bye!
INFO opsdroid.core: Exiting application with return code 0.
```

# Checklist:

- [ X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: Dockerfile
Comment (jacobtomlinson) [inline Dockerfile:2]: I'd prefer to keep me as the labelled maintainer of these images.

File: Dockerfile
Comment (jacobtomlinson) [inline Dockerfile:55]: Can we call this user `opsdroid`?

File: Dockerfile
Comment (jacobtomlinson) [inline Dockerfile:38]: Same here

File: Dockerfile
Comment (jacobtomlinson) [inline Dockerfile:53]: This version number looks funny. Is versioneer not working properly still?

**Review by jacobtomlinson — COMMENTED** [review]
I'm generally in favour of this, thanks for the efforts @rgreinho.

I would be keen to have a review from @Cadair given the Matrix dependencies are some of the most complex here.

File: Dockerfile
Comment (rgreinho) [reply Dockerfile:55]: Addressed ✅

File: Dockerfile
Comment (rgreinho) [reply Dockerfile:38]: Addressed ✅

File: Dockerfile
Comment (rgreinho) [reply Dockerfile:53]: I think it is just because there is no git repository within the container. 

If I run it locally I get something like`opsdroid-0.23.0+1.g6d81c23-py3-none-any.whl`.

Comment (jacobtomlinson) [reply Dockerfile:55]: Sure, but we want to fix this (see #950).

Switching to multi-stage builds here is a nice opportunity to resolve this problem because you can pull in the whole repo into the build stage and get the correct version in the package.

Comment (rgreinho) [reply Dockerfile:55]: This is not really relevant for the docker build though, as the package name as no impact on what is built. What would the benefit be? 

#950 seems to be a non-issue and should be closed.

Comment (jacobtomlinson) [reply Dockerfile:55]: I disagree. The `opsdroid.__version__` variable is also set by versioneer at build time and is used in multiple places. #950 mainly covers ensuring that variable is correct.

In the docker images this gets set incorrectly currently, hence my request to fix it in this PR as part of the awesome work you're doing here.

**Review by Dissonant-Tech — CHANGES_REQUESTED** [review]
@rgreinho Try this out. Should fix #950

File: Dockerfile
Comment (Cadair) [inline Dockerfile:52]: I think we will still need `olm-dev` here as it's a runtime dep for the e2ee matrix stuff as well as a compile time one.

Comment (Cadair) [reply Dockerfile:52]: Although presumably we might only need `olm` not `olm-dev`?

File: setup.cfg
Comment (Cadair) [inline setup.cfg:138]: Let's just delete this whole block?

File: Dockerfile
Comment (Cadair) [inline Dockerfile:62]: This doesn't interfere with running the container with `--user` right?

Comment (jacobtomlinson) [reply Dockerfile:62]: Nah it just changes the default user.

File: Dockerfile
Comment (rgreinho) [reply Dockerfile:55]: Oh this could simply be achieved using a wildcard directly:

```python
${DEPS_DIR}/opsdroid-*-py3-none-any.whl${EXTRAS}
```

However that does not fix the bug in #950 because `versioneer` is still unable to determine the version number automatically.

**Comment by jacobtomlinson** [comment]
Superseded by #1820

**Pull request closed without merging**
